### DISCUSSION TITLE
Intermediate Matrices

### DISCUSSION THREAD

**Discussion by ekwan** [grimme-lab/xtb#478] [Q&A]
Hi! Is it possible to gain access to any of the intermediate matrices of the calculation like the density matrix, overlap matrix, etc.? Thanks.

**Comment by awvwgk** [comment]
Not in the current infrastructure. What are you planning to do?

**Reply by sarmaier** [reply]
Hi! I wondered if the infrastructure has been updated to print these matrices? Thanks!

**Reply by awvwgk** [reply]
It's an open source project, contributing a patch might make this feature available sooner.

**Comment by ekwan** [comment]
Well there are all kinds of interesting structural analyses that can be done with that information, which is why many programs have some way to print it. Would you be able to suggest a way to get at this information, even if it means digging into the internals?

**Comment by awvwgk** [comment]
main/property.F90 might be a good place to start (it is always called after the last singlepoint and reevaluates the integrals):

https://github.com/grimme-lab/xtb/blob/e5c5bef7e5b1162a99a623676313efed41398df0/src/main/property.F90#L135-L145

`wfx` contains the density matrix (`wfx%P`) and the coefficients (`wfx%C`), as well as eigenvalues and occupation numbers and all the other important information. So there should be everything available you are looking for, you just have to decide on a format to print it. The ordering of the d orbitals is a bit strange (2, 0, -2, 1, -1; if I remember correctly).

**Reply by sarmaier** [reply]
I was able to find everything I need in main/property.F90. Thank you so much for the insight!

**Reply by dpadula85** [reply]
did you write some additional code to get some of these quantities? I'd be interested too!

**Reply by sarmaier** [reply]
I did. I just wrote a subroutine to print the matrices I wanted. I'm considering contributing a patch.
Because the subroutine "print_mulliken" in property.F90 is called every time for xtb1 and "print_wiberg" is called every time for xtb2, I suppose you could just write a print statement in one of those if you didn't care about printing the matrices every time.

**Reply by kvkarandashev** [reply]
Sorry to be a bother, but in the end have you contributed this as a patch? Getting more of xTB's intermediate data is important for me, so I may need something similar to what you did in the near future.
(Even though I think I might even need more data than in your case, as ideally I would prefer to have overlap matrix printed along with both canonical and localized orbitals.)

**Reply by sarmaier** [reply]
Hi, I unfortunately have not yet done this. If I do it soon, I'll let you know. Although not the best alternative, are you opposed to writing a print statement as I suggested above?

**Reply by kvkarandashev** [reply]
Hello! If by the time the issue becomes pressing for me the patch is not contributed I will indeed just add a print statement; the reason I consider this undesirable is that I do not want someone potentially having trouble reproducing my results due to them using a specifically modified version of code.

**Reply by awvwgk** [reply]
Would something like this work?

[![asciicast](https://asciinema.org/a/kjADav7aQyLYbHTQhFGojOwJN.svg)](https://asciinema.org/a/kjADav7aQyLYbHTQhFGojOwJN)

**Reply by awvwgk** [reply]
Since the calculator is stateless, it returns an object with all results of the calculation, which can be queried for specific properties or turned into a dict with all calculated properties. See https://github.com/tblite/tblite/pull/44 for the implementation, manylinux wheels for testing are available at my fork: https://github.com/awvwgk/tblite/actions/runs/2357386763. Let me if this works as expected.

**Reply by kvkarandashev** [reply]
Thank you very much! I think it does cover all my needs; should I have any questions or problems I will file them in tblite's "Issues" section.

**Reply by sarmaier** [reply]
Thanks for the awesome solution!

**Reply by awvwgk** [reply]
You're welcome.

The new library is still in early development, but I'm confident that we can put it to use in `xtb` soon as it is already used in DFTB+ and NWChem. Contributions to [*tblite*](https://github.com/tblite/tblite) are very welcome, the main library is well-documented, modern Fortran, but I'm also looking for help with improving the functionality of the Python bindings.

**Reply by linminhtoo** [reply]
hi, sorry to bump this after many months. is there ongoing work to merge this awesome contribution into `xtb` itself? (on top of having a separate project `tblite`) i would like to use `xtb` to print those intermediate matrices but i've never coded in Fortran so i'm not sure how difficult it is to manually add some print statements

**Reply by linminhtoo** [reply]
update: i managed to install `tblite` and get it working. it is a fantastic wrapper over `xtb` and i managed to output the intermediate matrices.

however i am also in need of the final fock matrix `F` which `tblite` doesn't seem to output in the results dictionary (created an Issue in that repo as well: https://github.com/tblite/tblite/issues/101) 

of course, there might also be a way to mathematically derive `F` based on the other matrices... but from my naive understanding we'd need additional info, like the energy eigenvalue matrix epsilon (from F.C = S.C.\epsilon)or the 2-electron electron repulsion integral tensor I. 

any help on this would be highly highly appreciated! thank you
